File: my-tinycards-api/app/models/deck.rb
class Deck < ApplicationRecord
  belongs_to :user

  has_many :cards, dependent: :destroy
  has_many :lessons, dependent: :destroy
  has_many :favorites, dependent: :destroy

  validates :title, presence: true, length: {maximum: 100}
  validates :description, presence: true, length: {maximum: 250}

  accepts_nested_attributes_for :cards

  scope :favorite_by, -> user {
    joins(:favorites).where(favorites: {user: user})
  }
  scope :by_following_people_of, -> user {
    where user: user.following
  }
  scope :trending, -> user {
    left_joins(:favorites).group(:id).order("COUNT(favorites.id) DESC")
  }

  delegate :name, to: :user, prefix: true, allow_nil: true
  
  ATTRIBUTE_PARAMS = [
  	:user_id,
    :title,
    :description,
    :cover_image,
    cards_attributes: [
      :id,
      :front,
      :back,
      :picture
    ]
  ].freeze

  mount_base64_uploader :cover_image, DeckUploader
end

File: angular2/src/app/deck.ts
import { Card } from './card'
import { Lesson } from './lesson'

export interface Deck {
	id: number,
	user_id: number,
  title: string,
	user_name: string,
	description: string,
  cover_image: string,
  favorite_count: number,
  favorited: boolean,
  cards: Card[],
  lessons: Lesson[],
}
